### DISCUSSION TITLE
[FAIL: panic: division or modulo by zero (0x12)]

### DISCUSSION THREAD

**Discussion by muhammadjehanzaib** [Cyfrin/foundry-full-course-cu#2782] [General]
forge test --mt testCanDepositedCollateralAndGetAccountInfo -vvvv
[⠒] Compiling...
No files changed, compilation skipped

Ran 1 test for test/unit/DecentralizedStableCoin.t.sol:TestDSC
[FAIL: panic: division or modulo by zero (0x12)] testCanDepositedCollateralAndGetAccountInfo() (gas: 121377)
Traces:
  [9520759] TestDSC::setUp()
    ├─ [4255594] → new DeployDSC@0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f
    │   └─ ← [Return] 21143 bytes of code
    ├─ [5008611] DeployDSC::run()
    │   ├─ [3299777] → new <unknown>@0x104fBc016F4bb334D775a19E8A6510109AC63E00
    │   │   ├─ [0] VM::startBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   ├─ [372255] → new <unknown>@0x90193C961A926261B756D1E5bb255e67ff9498A1
    │   │   │   └─ ← [Return] 1082 bytes of code
    │   │   ├─ [516505] → new <unknown>@0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496
    │   │   │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: DeployDSC: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f], value: 2000000000000000000000 [2e21])
    │   │   │   └─ ← [Return] 2117 bytes of code
    │   │   ├─ [372255] → new <unknown>@0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3
    │   │   │   └─ ← [Return] 1082 bytes of code
    │   │   ├─ [516505] → new <unknown>@0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76
    │   │   │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: DeployDSC: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f], value: 1000000000000000000000 [1e21])
    │   │   │   └─ ← [Return] 2117 bytes of code
    │   │   ├─ [0] VM::stopBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   └─ ← [Return] 6163 bytes of code
    │   ├─ [912] 0x104fBc016F4bb334D775a19E8A6510109AC63E00::activeNetworkConfig() [staticcall]
    │   │   └─ ← [Return] 0x90193C961A926261B756D1E5bb255e67ff9498A1, 0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3, 0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496, 0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76, 77814517325470205911140941194401928579557062014761831930645393041380819009408 [7.781e76]
    │   ├─ [0] VM::startBroadcast(<pk>)
    │   │   └─ ← [Return] 
    │   ├─ [591255] → new <unknown>@0x5FbDB2315678afecb367f032d93F642f64180aa3
    │   │   ├─ emit OwnershipTransferred(previousOwner: 0x0000000000000000000000000000000000000000, newOwner: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266)
    │   │   └─ ← [Return] 2609 bytes of code
    │   ├─ [832626] → new DSCEngine@0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512
    │   │   └─ ← [Return] 3481 bytes of code
    │   ├─ [2424] 0x5FbDB2315678afecb367f032d93F642f64180aa3::transferOwnership(DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512])
    │   │   ├─ emit OwnershipTransferred(previousOwner: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266, newOwner: DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512])
    │   │   └─ ← [Stop] 
    │   ├─ [0] VM::stopBroadcast()
    │   │   └─ ← [Return] 
    │   └─ ← [Return] 0x5FbDB2315678afecb367f032d93F642f64180aa3, DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], 0x104fBc016F4bb334D775a19E8A6510109AC63E00
    ├─ [912] 0x104fBc016F4bb334D775a19E8A6510109AC63E00::activeNetworkConfig() [staticcall]
    │   └─ ← [Return] 0x90193C961A926261B756D1E5bb255e67ff9498A1, 0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3, 0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496, 0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76, 77814517325470205911140941194401928579557062014761831930645393041380819009408 [7.781e76]
    ├─ [390] 0x5FbDB2315678afecb367f032d93F642f64180aa3::owner() [staticcall]
    │   └─ ← [Return] DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512]
    ├─ [24912] 0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496::mint(DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], 10000000000000000000 [1e19])
    │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], value: 10000000000000000000 [1e19])
    │   └─ ← [Stop] 
    ├─ [24912] 0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76::mint(DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], 10000000000000000000 [1e19])
    │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], value: 10000000000000000000 [1e19])
    │   └─ ← [Stop] 
    └─ ← [Stop] 

  [121377] TestDSC::testCanDepositedCollateralAndGetAccountInfo()
    ├─ [0] VM::startPrank(DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512])
    │   └─ ← [Return] 
    ├─ [24739] 0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496::approve(DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], 10000000000000000000 [1e19])
    │   ├─ emit Approval(owner: DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], spender: DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], value: 10000000000000000000 [1e19])
    │   └─ ← [Return] true
    ├─ [81067] DSCEngine::depositCollateral(0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496, 10000000000000000000 [1e19])
    │   ├─ emit CollateralDeposited(user: DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], token: 0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496, amount: 10000000000000000000 [1e19])
    │   ├─ [8988] 0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496::transferFrom(DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], 10000000000000000000 [1e19])
    │   │   ├─ emit Transfer(from: DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], to: DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], value: 10000000000000000000 [1e19])
    │   │   └─ ← [Return] true
    │   ├─ [8993] 0x90193C961A926261B756D1E5bb255e67ff9498A1::latestRoundData() [staticcall]
    │   │   └─ ← [Return] 1, 200000000000 [2e11], 1, 1, 1
    │   ├─ [8993] 0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3::latestRoundData() [staticcall]
    │   │   └─ ← [Return] 1, 100000000000 [1e11], 1, 1, 1
    │   └─ ← [Revert] panic: division or modulo by zero (0x12)
    └─ ← [Revert] panic: division or modulo by zero (0x12)

Suite result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 2.89ms (266.06µs CPU time)

Ran 1 test suite in 2.21s (2.89ms CPU time): 0 tests passed, 1 failed, 0 skipped (1 total tests)

Failing tests:
Encountered 1 failing test in test/unit/DecentralizedStableCoin.t.sol:TestDSC
[FAIL: panic: division or modulo by zero (0x12)] testCanDepositedCollateralAndGetAccountInfo() (gas: 121377)

Encountered a total of 1 failing tests, 0 tests succeeded

**Comment by muhammadjehanzaib** [comment]
//SPDX-License-Identifier: MIT
pragma solidity ^0.8.19;

import {Test, console} from "forge-std/Test.sol";
import {DeployDSC} from "../../script/DeployDSC.s.sol";
import {DecentralizedStableCoin} from "../../src/DecentralizedStableCoin.sol";
import {DSCEngine} from "../../src/DSCEngine.sol";
import {HelperConfig} from "../../script/HelperConfig.s.sol";
import {ERC20Mock} from "../Mocks/ERC20Mock.sol";

contract TestDSC is Test {
    // error DSCEngine__NeedMoreThanZero();
    DecentralizedStableCoin decentralizedStableCoin;
    DSCEngine dscEngine;
    DeployDSC deployDSC;
    HelperConfig config;
    address ethUSDPriceFee;
    address btcUSDPriceFee;
    address weth;
    address wbtc;
    address public fieldPlayer = makeAddr("fieldPlayer");
    uint256 constant public AMOUNT_COLLATERAL = 10 ether;

    function setUp() external {
        deployDSC = new DeployDSC();
        // vm.prank(fieldPlayer);
        (decentralizedStableCoin, dscEngine, config) = deployDSC.run();
        (ethUSDPriceFee,btcUSDPriceFee, weth, wbtc,) = config.activeNetworkConfig();
        fieldPlayer = decentralizedStableCoin.owner();
        ERC20Mock(weth).mint(fieldPlayer,  AMOUNT_COLLATERAL);
        ERC20Mock(wbtc).mint(fieldPlayer,  AMOUNT_COLLATERAL);
    }

    ///////////////////////////////////
    // constructor Tests //////////////
    ///////////////////////////////////
    address[] public tokenAddress;
    address[] public priceFeedAddress;

    function testRevertsIfTokenLengthDoesnotMactchPriceFee () public{
        tokenAddress.push(weth);
        priceFeedAddress.push(ethUSDPriceFee);
        priceFeedAddress.push(ethUSDPriceFee);

        vm.expectRevert(DSCEngine.DSCEngine__TokenAddressAndPriceFeedAddressesLengthMustBeSame.selector);
        new DSCEngine(tokenAddress, priceFeedAddress,address(decentralizedStableCoin));
    }

    function testGetTokenAmountFormUSD() public view{
        uint256 usdAmount = 100 ether;
        uint256 expectedWeth = 0.05 ether;
        // 2000 / 100
        uint256 pirceInWei = dscEngine.getTokenAmountFromUsd(weth, usdAmount);

        assertEq(expectedWeth,pirceInWei);

    }

    ///////////////////////////////////
    // DSC Engin Tests ////////////////
    ///////////////////////////////////

    function testGetUsedValue() public view {
        uint256 ethAmount = 15e18;
        uint256 expectedUsd = 30000e18;
        uint256 actualUsd = dscEngine.getUsdValue(weth, ethAmount);

        assertEq(expectedUsd, actualUsd);
    }

    function testRevertIfCollateralZero() public {
        vm.startPrank(fieldPlayer);
        ERC20Mock(weth).approveInternal(address(dscEngine), fieldPlayer, 10 ether);

        vm.expectRevert(DSCEngine.DSCEngine__NeedMoreThanZero.selector);
        dscEngine.depositCollateral(weth, 0);
        vm.stopPrank();
    }


    function testRevertWithUnapprovedCollateral() public {
        ERC20Mock ranToken = new ERC20Mock("Random Token","RAN", fieldPlayer, 1002 ether);
        vm.startPrank(fieldPlayer);
        vm.expectRevert(DSCEngine.DSCEngine__NotAllowedToken.selector);
        dscEngine.depositCollateral(address(ranToken), 10 ether);
        vm.stopPrank();
    }

    modifier depositedCollateral() {
        vm.startPrank(fieldPlayer);
        ERC20Mock(weth).approve(address(dscEngine),AMOUNT_COLLATERAL);
        dscEngine.depositCollateral(weth, AMOUNT_COLLATERAL);
        vm.stopPrank();
        _;
    }

    // function testCanDepositCollateralAndAccountInfo() public depositedCollateral{
    //     (uint256 totalDSCMinted, uint256 collateralValueInUsd)  = dscEngine.getAccontInformation(fieldPlayer);

    //     uint256 expectectedTotalDscMinted = 0;
    //     uint256 expectedDepositedAmmount= dscEngine.getTokenAmountFromUsd(weth, collateralValueInUsd);
    //     assertEq(totalDSCMinted, expectectedTotalDscMinted);
    //     assertEq(AMOUNT_COLLATERAL, expectedDepositedAmmount);
    // }
    function testCanDepositedCollateralAndGetAccountInfo() public depositedCollateral {
        (uint256 totalDscMinted, uint256 collateralValueInUsd) = dscEngine.getAccontInformation(fieldPlayer);
        uint256 expectedDepositedAmount = dscEngine.getTokenAmountFromUsd(weth, collateralValueInUsd);
        assertEq(totalDscMinted, 0);
        assertEq(expectedDepositedAmount, AMOUNT_COLLATERAL);
    }

    ///////////////////////////////////
    //Decentralized Stable Coin Tests//
    ///////////////////////////////////
    function testAmountZeroRevert() public {
        vm.expectRevert();
        decentralizedStableCoin.burn(0);
    }

    function testBalanceAmountIsLessThanExpectedAmount() public {
        console.log(fieldPlayer);
        vm.prank(fieldPlayer);
        decentralizedStableCoin.mint(fieldPlayer, 3000000000000000000);
        vm.expectRevert();
        decentralizedStableCoin.burn(2000000000000000000);
    }

    function testMintingTokens() public {
        vm.prank(fieldPlayer);
        bool isMinted = decentralizedStableCoin.mint(fieldPlayer, 2000e18);
        assertEq(2000e18, decentralizedStableCoin.balanceOf(fieldPlayer));
        assert(isMinted);
    }

    function testmintAmountShouldGreaterThanZero() public {
        vm.prank(fieldPlayer);
        vm.expectRevert();
        decentralizedStableCoin.mint(address(this), 0);
    }

    function testAddressShouldNotBeZero() public {
        vm.prank(fieldPlayer);
        vm.expectRevert();
        decentralizedStableCoin.mint(address(0), 1000000);
    }

    function testBurningTokens() public {
        vm.prank(fieldPlayer);
        decentralizedStableCoin.mint(fieldPlayer, 2000e18);
        vm.prank(fieldPlayer);
        decentralizedStableCoin.burn(1000e18);
        assertEq(decentralizedStableCoin.balanceOf(fieldPlayer), 1000e18);
    }
}

**Comment by EngrPips** [comment]
Hello @muhammadjehanzaib, Please show the `healthFactor` function in your `DSCEngine` contract.

**Comment by muhammadjehanzaib** [comment]
//SPDX-License-Identifier: MIT
pragma solidity ^0.8.19;

import {DecentralizedStableCoin} from "../src/DecentralizedStableCoin.sol";
import {ReentrancyGuard} from "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";

/**
 * @title DSCEngine
 */
contract DSCEngine is ReentrancyGuard {
    error DSCEngine__NeedMoreThanZero();
    error DSCEngine__TokenAddressAndPriceFeedAddressesLengthMustBeSame();
    error DSCEngine__NotAllowedToken();
    error DSCEngine__TransferFailed();
    error DSCEngine__HealthFactorIsBelowMinimum(uint256 healthFactor);
    error DSCEngine__MintFailed();
    error DSCEngine__BreaksHealthFactorOk();
    error DSCEngine__HealthFactorNotImproved();

    uint256 private constant ADDITIONAL_FEE_PRECISION = 1e10;
    uint256 private constant PRECISION = 1e18;
    uint256 private constant LIQUIDATION_THRESHOLD = 50;
    uint256 private constant LIQUIDATION_PRECISION = 100;
    uint256 private constant MIN_HEALTH_FACTOR = 1e18;
    uint256 private constant LIQUIDATION_BONUS = 10;

    mapping(address token => address priceFeed) private s_priceFeeds;
    mapping(address user => mapping(address token => uint256 amount)) private s_collateralDeposite;
    mapping(address user => uint256 amountDscToMinted) private s_DSCMinted;
    address[] private s_collateralTokens;
    DecentralizedStableCoin private immutable i_DSCADDRESS;

    event CollateralDeposited(address indexed user, address indexed token, uint256 indexed amount);
    event CollateralRedeemed(
        address indexed redeemedFrom, address redeemedTo, address indexed token, uint256 indexed amount
    );

    modifier moreThanZero(uint256 amount) {
        if (amount == 0) {
            revert DSCEngine__NeedMoreThanZero();
        }
        _;
    }

    modifier isAllowedToken(address token) {
        if (s_priceFeeds[token] == address(0)) {
            revert DSCEngine__NotAllowedToken();
        }
        _;
    }

    constructor(address[] memory tokenAddresses, address[] memory priceFeedAddresses, address dscAddress) {
        if (tokenAddresses.length != priceFeedAddresses.length) {
            revert DSCEngine__TokenAddressAndPriceFeedAddressesLengthMustBeSame();
        }
        for (uint256 i = 0; i < tokenAddresses.length; i++) {
            s_priceFeeds[tokenAddresses[i]] = priceFeedAddresses[i];
            s_collateralTokens.push(tokenAddresses[i]);
        }
        i_DSCADDRESS = DecentralizedStableCoin(dscAddress);
    }

    function depositCollateralAndMintDsc(
        address tokenCollateralAddress,
        uint256 amountCollateral,
        uint256 amountDSCToMint
    ) public {
        depositCollateral(tokenCollateralAddress, amountCollateral);
        mintDsc(amountDSCToMint);
    }

    /**
     * @param tokenCollateralAddress The address of token to deposit as collateral
     * @param amountCollateral The Total collateral amount want to deposit
     */
    function depositCollateral(address tokenCollateralAddress, uint256 amountCollateral)
        public
        moreThanZero(amountCollateral)
        isAllowedToken(tokenCollateralAddress)
        nonReentrant
    {
        s_collateralDeposite[msg.sender][tokenCollateralAddress] += amountCollateral;
        emit CollateralDeposited(msg.sender, tokenCollateralAddress, amountCollateral);
        bool success = IERC20(tokenCollateralAddress).transferFrom(msg.sender, address(this), amountCollateral);
        if (!success) {
            revert DSCEngine__TransferFailed();
        }
        _revertIfHealthFactorIsBroken(msg.sender);
    }

    function redeemCollateralForDsc(address tokenCollateralAddress, uint256 amountCollateral, uint256 amountDscToBurn)
        external
    {
        burnDsc(amountDscToBurn);
        redeemCollateral(tokenCollateralAddress, amountCollateral);
    }

    function redeemCollateral(address tokenCollateralAddress, uint256 amountCollateral)
        public
        moreThanZero(amountCollateral)
        nonReentrant
    {
        _redeemCollateral(tokenCollateralAddress, amountCollateral, msg.sender, msg.sender);
        _revertIfHealthFactorIsBroken(msg.sender);
    }

    function mintDsc(uint256 amountDscToMint) public moreThanZero(amountDscToMint) nonReentrant {
        s_DSCMinted[msg.sender] += amountDscToMint;
        // if they minted too much ($150 DSC, $100)

        _revertIfHealthFactorIsBroken(msg.sender);
        bool minted = i_DSCADDRESS.mint(msg.sender, amountDscToMint);
        if (!minted) {
            revert DSCEngine__MintFailed();
        }
    }

    function burnDsc(uint256 amount) public moreThanZero(amount) {
        _burnDsc(amount,msg.sender, msg.sender);
        _revertIfHealthFactorIsBroken(msg.sender);
    }

    function liquidate(address collateral, address user, uint256 debtToCover)
        external
        moreThanZero(debtToCover)
        nonReentrant
    {
        uint256 startingUserHealthFactor = _healthFactor(user);
        if (startingUserHealthFactor >= MIN_HEALTH_FACTOR) {
            revert DSCEngine__BreaksHealthFactorOk();
        }
        uint256 tokenAmountFromDebtCovered = getTokenAmountFromUsd(collateral, debtToCover);
        uint256 bonusCollateral = (tokenAmountFromDebtCovered * LIQUIDATION_BONUS) / LIQUIDATION_PRECISION;
        uint256 totalCollateralToRedeem = tokenAmountFromDebtCovered + bonusCollateral;
        _redeemCollateral(collateral, totalCollateralToRedeem, user, msg.sender);
        _burnDsc(debtToCover, user, msg.sender);

        uint256 endingUserHealthFactor = _healthFactor(user);
        if (endingUserHealthFactor <= startingUserHealthFactor) {
            revert DSCEngine__HealthFactorNotImproved();
        }
        _revertIfHealthFactorIsBroken(msg.sender);
    }

    function getTokenAmountFromUsd(address token, uint256 amountInWei) public view returns (uint256) {
        AggregatorV3Interface priceFeed = AggregatorV3Interface(s_priceFeeds[token]);
        (, int256 price,,,) = priceFeed.latestRoundData();
        return (amountInWei * PRECISION) / (uint256(price) * ADDITIONAL_FEE_PRECISION);
    }

    function _getAccountInformation(address user)
        private
        view
        returns (uint256 totalDSCMinted, uint256 collateralInValueUsd)
    {
        totalDSCMinted = s_DSCMinted[user];
        collateralInValueUsd = getAccountCollateralValue(user);
    }

    function _healthFactor(address user) private view returns (uint256) {
        // total DSC minted
        // total collateral VALUE
        (uint256 totalDSCMinted, uint256 collateralInValueUsd) = _getAccountInformation(user);
        uint256 collateralAdjustedForThreshold = (collateralInValueUsd * LIQUIDATION_THRESHOLD) / LIQUIDATION_PRECISION;
        //
        return (collateralAdjustedForThreshold * PRECISION) / totalDSCMinted;
    }

    function _burnDsc(uint256 amountDscToBurn, address onBehalfOf, address dscFrom) private {
        s_DSCMinted[onBehalfOf] -= amountDscToBurn;
        bool success = i_DSCADDRESS.transferFrom(dscFrom, address(this), amountDscToBurn);
        if (!success) {
            revert DSCEngine__TransferFailed();
        }
        i_DSCADDRESS.burn(amountDscToBurn);
    }

    function _redeemCollateral(address tokenCollateralAddress, uint256 amountCollateral, address from, address to)
        private
    {
        s_collateralDeposite[from][tokenCollateralAddress] -= amountCollateral;
        emit CollateralRedeemed(from, to, tokenCollateralAddress, amountCollateral);

        bool success = IERC20(tokenCollateralAddress).transfer(to, amountCollateral);
        if (!success) {
            revert DSCEngine__TransferFailed();
        }
    }

    function _revertIfHealthFactorIsBroken(address user) internal view {
        uint256 userHealthFactor = _healthFactor(user);

        if (userHealthFactor < MIN_HEALTH_FACTOR) {
            revert DSCEngine__HealthFactorIsBelowMinimum(userHealthFactor);
        }
    }

    function getAccountCollateralValue(address user) public view returns (uint256 totalCollateralValueInUsd) {
        for (uint256 i = 0; i < s_collateralTokens.length; i++) {
            address token = s_collateralTokens[i];
            uint256 amount = s_collateralDeposite[user][token];
            totalCollateralValueInUsd += getUsdValue(token, amount);
        }
        return totalCollateralValueInUsd;
    }

    function getUsdValue(address token, uint256 amount) public view returns (uint256) {
        AggregatorV3Interface priceFee = AggregatorV3Interface(s_priceFeeds[token]);
        (, int256 price,,,) = priceFee.latestRoundData();
        // 1 Eth = $1000
        // we will get in return 1000 * 1e8 (Becuase usdt has 8 decimals)
        // we will explicitly add 10(1e10) more decimals here in order to get exact value
        // after it all we will divide the whole equation  by 1e18 in order to get value in decimals
        return ((uint256(price) * ADDITIONAL_FEE_PRECISION) * amount) / PRECISION;
    }

    function getAccontInformation(address user) public  view returns (uint256, uint256) {
        (uint256 totalDscMinted, uint256 totalValueInUSD) =  _getAccountInformation(user);
        return  (totalDscMinted, totalValueInUSD);
    }
}

**Comment by EngrPips** [comment]
Please have your `_healthFactor` as the below then retry

```js
function _healthFactor(address user) private view returns (uint256) {
// total DSC minted
// total collateral VALUE
(uint256 totalDSCMinted, uint256 collateralInValueUsd) = _getAccountInformation(user);
if(totalDSCMinted == 0) return type(uint256).max;
uint256 collateralAdjustedForThreshold = (collateralInValueUsd * LIQUIDATION_THRESHOLD) / LIQUIDATION_PRECISION;
//
return (collateralAdjustedForThreshold * PRECISION) / totalDSCMinted;
}
```

**Comment by muhammadjehanzaib** [comment]
it works, can you explain this line please and why we adding this here ...
'''solidity
if(totalDSCMinted == 0) return type(uint256).max;
'''

**Reply by EngrPips** [reply]
So without this line of code `if(totalDSCMinted == 0) return type(uint256).max;` the `totalDSCMinted` will be `0` and then we will use it to divide `(collateralAdjustedForThreshold * PRECISION)` in this line of code `return (collateralAdjustedForThreshold * PRECISION) / totalDSCMinted;` and by doing so the `evm` will revert because dividing by `0` cause a panic error in solidity. Please ask questions if anything is confusing.

**Reply by muhammadjehanzaib** [reply]
All set ... Thanks for your time!

**Reply by EngrPips** [reply]
You are welcome, Keep crushing it.
